### PR TITLE
Fix editing the message

### PR DESCRIPTION
**Pull request by Piotrfj**
### Fixed Issues

$ https://github.com/Expensify/App/issues/18805
PROPOSAL: https://github.com/Expensify/App/issues/18805#issuecomment-1636227437

### Tests

- [x] Verify that no errors appear in the JS console
Scenario 1:
1. Open chat as a user A with user B.
2. write a message and open edit mode on it.
3. As a user B write couple (or just spam a lot) of messages on the same chat.
4. The edit box should not be pushed up as new messages are appearing.

Scenario 2:
1. Open chat and open edit mode on some report.
2. Lose focus on the edit box and move it somewhere upper on the screen.
3. Focus the box again, the box should be moved to the bottom of the screen without any troubles.

### Offline tests

Bug is only testable on online mode since we are testing behavior when new messages are appearing.

### QA Steps

- [ ] Verify that no errors appear in the JS console

Scenario 1:
1. Open chat as a user A with user B.
2. write a message and open edit mode on it.
3. As a user B write couple (or just spam a lot) of messages on the same chat.
4. The edit box should not be pushed up as new messages are appearing.

Scenario 2:
1. Open chat and open edit mode on some report.
2. Lose focus on the edit box and move it somewhere upper on the screen.
3. Focus the box again, the box should be moved to the bottom of the screen without any troubles.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/41384305/4845ed06-8815-4348-b235-e4337401fe6f

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/41384305/1452cb69-38bc-4092-8d24-77b524ff5ae8

</details>

### CODE REVIEW

File: src/pages/home/report/ReportActionListFrozenScrollContext.js
Comment (rezkiy37) [inline src/pages/home/report/ReportActionListFrozenScrollContext.js:5]: Let's keep value and setter in one place.
```suggestion
const ReportActionListFrozenScrollContext = React.createContext({
    isScrollFrozen: false,
    setShouldFreezeScroll: () => {},
});
```

File: src/pages/home/report/ReportActionsList.js
Comment (rezkiy37) [inline src/pages/home/report/ReportActionsList.js:176]: I guess we should wrap entire **ReportScreen**. I believe it could help in the future to get the value and setter somewhere within the screen.

**Review by rezkiy37 — COMMENTED** [review]
Left a couple comments

File: src/pages/home/report/ReportActionsList.js
Comment (gedu) [reply src/pages/home/report/ReportActionsList.js:176]: In general, the approach is to create a [HOC](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#composition-vs-inheritance), which isn't clean to throw the Provider in the middle of a component. 
Also if possible, given that it is a function component, maybe a hook would work.

I would simply wrap this component with the hook, I wouldn't add it at the `ReportScreen` level, given that if not needed, any change will rerender all the children's components

File: src/pages/home/report/ReportActionsList.js
Comment (gedu) [inline src/pages/home/report/ReportActionsList.js:217]: this object can be created outside of the component

File: src/pages/home/report/ReportActionListFrozenScrollContext.js
Comment (rezkiy37) [inline src/pages/home/report/ReportActionListFrozenScrollContext.js:11]:
```suggestion
    shouldFreezeScroll: PropTypes.bool.isRequired,

    /** Function to update the state */
    setShouldFreezeScroll: PropTypes.func.isRequired,
```

File: src/pages/home/report/ReportActionListFrozenScrollContext.js
Comment (rezkiy37) [inline src/pages/home/report/ReportActionListFrozenScrollContext.js:33]: The `useMemo` looks like an overhead here.
cc @fabioh8010 @gedu 
```suggestion
        value={{
            shouldFreezeScroll,
            setShouldFreezeScroll,
        }}>{props.children}</ReportActionListFrozenScrollContext.Provider>;
```

**Review by rezkiy37 — COMMENTED** [review]
Left a couple comments. In general, LGTM.

File: src/pages/home/report/ReportActionListFrozenScrollContext.js
Comment (fabioh8010) [reply src/pages/home/report/ReportActionListFrozenScrollContext.js:33]: I think it's correct @rezkiy37, because if we pass an object in this way, is like we are creating a new object every time and it will trigger unnecessary re-renders (old object will never equal newly created one). It makes sense to me to memoize.

File: src/pages/home/report/ReportActionListFrozenScrollContext.js
Comment (fabioh8010) [inline src/pages/home/report/ReportActionListFrozenScrollContext.js:42]: Why are we creating this HOC? It's not being used anywhere.

File: src/hooks/useFrozenScroll.js
Comment (ArekChr) [inline src/hooks/useFrozenScroll.js:9]: We could improve error handling here.

```suggestion
    const context = useContext(ReportActionListFrozenScrollContext);
    if (context === undefined)
        throw new Error('useFrozenScroll must be within ReportActionListFrozenScrollContextProvider');
    }
    return context;
```

File: src/pages/home/report/ReportActionListFrozenScrollContext.js
Comment (ArekChr) [reply src/pages/home/report/ReportActionListFrozenScrollContext.js:33]: This does not need to be memorized. We could return the `useState`object to the provider (as an array, not object), which will be the same reference. WDYT? 
```
function ReportActionListFrozenScrollContextProvider(props) {
    const shouldFreezeScrollState = useState(false);
    return (
        <ReportActionListFrozenScrollContext.Provider value={shouldFreezeScrollState}>
            {props.children}
        </ReportActionListFrozenScrollContext.Provider>
    );
}
```

Comment (rezkiy37) [reply src/pages/home/report/ReportActionListFrozenScrollContext.js:33]: Agree with @ArekChr approach.

**Review by Santhosh-Sellavel — CHANGES_REQUESTED** [review]
Resolve conflicts. Adding myself as a reviewer

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel conflicts are resolved

**Comment by Santhosh-Sellavel** [comment]
This ran into conflicts again.

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel looks like there was refactoring meanwhile to functional component, conflicts resolved

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj Please add recordings

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj Update what test/verify on different platforms.

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel added examples

**Comment by Santhosh-Sellavel** [comment]
Bug: iOS Animation is weird while entering edit mode.

The message being edited is at the top of the screen, animation should be like the edit composer scrolls down to the bottom.

Instead, it appears all of a sudden the view scrolls from the bottom.


https://github.com/Expensify/App/assets/85645967/70d4f3c6-e7cf-4d85-bc5d-43e2e8ebb159

**Comment by Santhosh-Sellavel** [comment]
I'm going on OOO for a week if PR is ready for a review. Please request for a new C+ on issue

**Comment by situchan** [comment]
I can continue review on this one

File: src/pages/home/ReportScreen.js
Comment (gedu) [inline src/pages/home/ReportScreen.js:121]: Fix styling, why all that spacing?

**Review by situchan — COMMENTED** [review]
Please mark as resolved on all addressed comments.
And fix conflicts as well

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel From what I understand you are complaining about the jumpy behavior of the scroll, that was intendent because animated scroll was causing issues relating with the bug. However, I have checked the scroll animation and after turning it back on nothing seems to break as it used to before, maybe its something related to recent changed in these files.
So after all, my sugestion is to bring them back, and leave the solution with only freezing the scroll during editting.

Video examples:

https://github.com/Expensify/App/assets/41384305/7968c910-ed00-4d62-82e9-a96ea8def7c1

**Comment by Santhosh-Sellavel** [comment]
Okay @Piotrfj Please resolve the conflicts, make PR ready for review again lets push this forward, thanks!

**Comment by marcochavezf** [comment]
Oops conflicts again

**Comment by Piotrfj** [comment]
@marcochavezf done :')

**Comment by marcochavezf** [comment]
Oh now lint errors 😌

**Comment by Piotrfj** [comment]
@marcochavezf done :)

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj Still lint failing!

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj 

Tests fails And Glitch while running test steps on a new thread/chat. Definitely Deploy blockers it could affect new users experience.

https://github.com/Expensify/App/assets/85645967/fdecb117-69c6-4d80-bca0-f70a8cc24c3d

**Review by Santhosh-Sellavel — CHANGES_REQUESTED** [review]
Bug: https://github.com/Expensify/App/pull/24154#issuecomment-1734374357

**Comment by gedu** [comment]
He is on sick leave, he will be back soon

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel looking at it today

**Comment by Santhosh-Sellavel** [comment]
Bump @Piotrfj

**Comment by Santhosh-Sellavel** [comment]
Bump @Piotrfj

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel There is a difference between safari and other environments in behavior I havent discovered before, currently Im working on the solution that would take care of that and will not affect other environments

**Comment by Santhosh-Sellavel** [comment]
I remember this affects all Mweb not just Safari, please check again & confirm.

**Comment by Santhosh-Sellavel** [comment]
Can resolve conflicts please?

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel just came back from sick leave, going back to work on the issue, will give some feedback today.

**Comment by Santhosh-Sellavel** [comment]
Any update @Piotrfj?

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel I have put the comment on different task by mistake, first solution was buggy and right now I'm close to complete second solution, but still I'm encountering problems with safari. Will give some more feedback tomorrow as I will discuss this with inner teammates.

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel After all this battle and consulting I haven't been able to provide well working solution. However we have found that the issue is in progres on react-native-web library. So we are suggesting to put our issue ON HOLD until the problem is resolved at source or merge current solution (only safari is not working) and open new issue just for safari.

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel https://github.com/necolas/react-native-web/pull/2588 thats the issue I was talking about.

**Comment by srikarparsi** [comment]
Hey! Please re-assign me when this PR is ready for review :)

**Comment by marcochavezf** [comment]
It's been a year without activity, I will close it out to remove it from the k2. However, feel free to re-open if there's something we need to look up

**Pull request closed without merging**
